File: 09_06_2024_Subarray_Sums_Divisible_by_K.py
class Solution(object):
    def subarraysDivByK(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        prefixMod, result = 0, 0

        modGroups = [0] * k
        modGroups[0] = 1

        for num in nums:
            prefixMod = (prefixMod + num) % k
            result += modGroups[prefixMod]
            modGroups[prefixMod] += 1

        return result

File: 09_06_2024_Subarray_Sums_Divisible_by_K.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int prefixMod = 0, result = 0;
        vector<int> modGroups(k, 0);
        modGroups[0] = 1;

        for (int num : nums) {
            prefixMod = (prefixMod + num) % k;
            if (prefixMod < 0) {
                prefixMod += k;
            }
            result += modGroups[prefixMod];
            modGroups[prefixMod]++;
        }

        return result;
    }
};

